### PR TITLE
Move hooks out of router

### PR DESCRIPTION
**Pull request by hswopeams**
PR for story "Move hooks out of BosonRouter into Cashier contract"

Changes to contracts, migration scripts (new references from contracts to other contracts and removal of references required changes to deployment scripts), and tests. Also added a new test file to test new functions such as `Cashier.setBosonRouterAddress`,which didn't have a clear place in any of the current test files. Also moved existing test cases that already set address references on contracts and removed duplicates.

_beforeERC1155Transfer was removed because it included several calls that in the end just access the `balances` mapping in the calling contract(ERC1155721.sol).

### CODE REVIEW

File: test/1_test_fullpath.js [hunk 1]
```diff
@@ -57,19 +53,31 @@ contract('Voucher tests', async (addresses) => {
     await contractERC1155ERC721.setVoucherKernelAddress(
       contractVoucherKernel.address
     );
-    await contractERC1155ERC721.setBosonRouterAddress(
-      contractBosonRouter.address
-    );
+
+    await contractERC1155ERC721.setCashierAddress(contractCashier.address);
+
+    await contractERC1155ERC721.setCashierAddress(contractCashier.address);
```
Comment (ali2251) [inline test/1_test_fullpath.js:59]: is it duplicated?

File: test/1_test_fullpath.js [hunk 2]
```diff
@@ -1148,19 +1151,32 @@ contract('Voucher tests - UNHAPPY PATH', async (addresses) => {
...
```
Comment (ali2251) [inline test/1_test_fullpath.js:1157]: same here, seems like its doing the same thing

File: test/7_ admin_functionality.js [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+
+    it('[NEGATIVE][setTokenContractAddress] Should revert if ZERO address is provided', async () => {
+      await truffleAssert.reverts(
+        contractCashier.setTokenContractAddress(constants.ZERO_ADDRESS),
+        truffleAssert.ErrorType.REVERT
+      );
+    });
+  });
+
+  describe('ERC1125721', () => {
```
Comment (ali2251) [inline test/7_ admin_functionality.js:129]: 1155 perhaps?

**Review by ali2251 — CHANGES_REQUESTED** [review]

File: contracts/Cashier.sol [hunk 4]
```diff
@@ -826,28 +841,158 @@ contract Cashier is ICashier, UsingHelpers, ReentrancyGuard, Ownable, Pausable {
...
+     * @param _to next owner of the _tokenIdVoucher
+     * @param _tokenIdVoucher _tokenIdVoucher that has been transferred
      */
-    function addEscrowTokensAmount(
-        address _token,
-        address _account,
-        uint256 _newAmount
-    ) external override onlyFromRouter {
-        escrowTokens[_token][_account] =  escrowTokens[_token][_account].add(_newAmount);
+    function onERC721Transfer(
```
Comment (HristiyanG) [inline contracts/Cashier.sol:858]: My understanding was that we move the hooks in a separate contract, responsible only for those validations? Why we need to extend the Cashier?

Comment (HristiyanG) [inline contracts/Cashier.sol:858]: My understanding was that we move the hooks in a separate contract, responsible only for those validations? Why we need to extend the Cashier?

File: contracts/Cashier.sol [hunk 5]
```diff
@@ -826,28 +841,158 @@ contract Cashier is ICashier, UsingHelpers, ReentrancyGuard, Ownable, Pausable {
...
     }
 
+    /**
+     * @notice After the transfer happens the _tokenSupplyId should be updated in the promise. Escrow funds for the seller's deposits (If in ETH) should be allocated to the new owner as well.
+     * @param _from prev owner of the _tokenSupplyId
+     * @param _to nex owner of the _tokenSupplyId
+     * @param _tokenSupplyId _tokenSupplyId for transfer
+     * @param _value qty which has been transferred
+     */
+    function onERC1155Transfer(
```
Comment (HristiyanG) [inline contracts/Cashier.sol:948]: Here as well

File: contracts/Cashier.sol [hunk 6]
```diff
@@ -826,28 +841,158 @@ contract Cashier is ICashier, UsingHelpers, ReentrancyGuard, Ownable, Pausable {
...
+                    _tokenSupplyId
+                );
+
+            depositSe = IVoucherKernel(voucherKernel).getSellerDeposit(
+                _tokenSupplyId
+            );
+            totalAmount = depositSe.mul(_value);
+
+            //Reduce _from escrow token amount and increase _to escrow token amount - deposit
+            escrowTokens[tokenDepositAddress][_from] =  escrowTokens[tokenDepositAddress][_from].sub(totalAmount);
```
Comment (HristiyanG) [inline contracts/Cashier.sol:985]: Two whitespaces after `=`
Please run the pipeline locally, so we avoid that.

File: contracts/Cashier.sol [hunk 7]
```diff
@@ -826,28 +841,158 @@ contract Cashier is ICashier, UsingHelpers, ReentrancyGuard, Ownable, Pausable {
...
+            escrowTokens[tokenDepositAddress][_to] =  escrowTokens[tokenDepositAddress][_to].add(totalAmount);
```
Comment (HristiyanG) [inline contracts/Cashier.sol:986]: Two whitespaces after `=`
Please run the pipeline locally, so we avoid that.

**Review by HristiyanG — CHANGES_REQUESTED** [review]

File: contracts/Cashier.sol (hunk 4 above)
Comment (hswopeams) [reply contracts/Cashier.sol:858]: The hooks were making a lot of calls to the Cashier. Some calls are being made to VouherKernel as well, but most of the external calls were made to Cashier. By moving the hooks to Cashier, the functions can directly acces `escrowTokens` and `escrow` mappings. It removes several external calls per hook function, which makes the hook functions more efficient

Comment (HristiyanG) [reply contracts/Cashier.sol:858]: Somewhat agree, but don't feel this is the correct place. You are correct we are making calls to Cashier and VK, but for me we should have dedicated contract responsible for the hooks. @ali2251 Do you have an opinion here?

File: contracts/Cashier.sol (hunk 6 above)
Comment (hswopeams) [reply contracts/Cashier.sol:985]: I did run the pipeline locally -- multiple times. I ran the `./go` script and also `./go contracts:lint`,` ./go contracts:lint_fix`, `./go contracts:format`, `./go contracts:format_fix`, `./go tests:lint`, `./go tests:lint_fix`,` ./go tests:format,` and `./go tests:format_fix` separately.

As you can see, it also passed when the pipeline was run from GitHub. This was never flagged. I will change it manually, but the scripts don't flag everything so some things can slip through.

File: contracts/Cashier.sol (hunk 5 above)
Comment (hswopeams) [reply contracts/Cashier.sol:948]: The hooks were making a lot of calls to the Cashier. Some calls are being made to VouherKernel as well, but most of the external calls were made to Cashier. By moving the hooks to Cashier, the functions can directly acces escrowTokens and escrow mappings. It removes several external calls per hook function, which makes the hook functions more efficient

File: contracts/Cashier.sol (hunk 7 above)
Comment (hswopeams) [reply contracts/Cashier.sol:986]: I did run the pipeline locally -- multiple times. I ran the ./go script and also ./go contracts:lint, ./go contracts:lint_fix, ./go contracts:format, ./go contracts:format_fix, ./go tests:lint, ./go tests:lint_fix, ./go tests:format, and ./go tests:format_fix separately.

As you can see, it also passed when the pipeline was run from GitHub. This was never flagged. I will change it manually, but the scripts don't flag everything so some things can slip through.

File: contracts/Cashier.sol (hunk 4 above)
Comment (ali2251) [reply contracts/Cashier.sol:858]: I think if we reach a point where we are out of space in Cashier, we should create a new contract and move these there but for now it seems the easiest to do it this way

**Review by HristiyanG — APPROVED** [review]

**Review by ali2251 — APPROVED** [review]
